### ISSUE TITLE
Pluggable transition mechanism

### ISSUE THREAD

**Issue by oliverhausler** [square/picasso#743]
I don't find it now, but I made a similar request a few months ago and you guys rejected it because you wanted to keep Picasso simple (a good thing, I agree; and my suggestion was not as clean as this one here). So let me bring this up again, please.

Suggestion: Add a fade() method, which always fades in an image, no matter if it comes from memory or the network.

Picasso.with(getContext()).load(url).into(viewContent); // implements your default behavior (fades in for network, does no fade in for memory)
Picasso.with(getContext()).load(url).noFade().into(viewContent); // never fades in (you have this already)
Picasso.with(getContext()).load(url).fade().into(viewContent); // always fades in (NEW) - you could provide an optional signature with an argument as well for the fade duration, if you wanted

Why is this useful? In image libraries, where many images are shown on the same screen, part of the images come from the network and part are in memory already, so the visual effect varies between images which is a no-go for any UI. Sure I can set .noFade() which makes the visuals the same no matter where the image comes from, but then the nice smooth fade-in is gone as well. Or I can control it manually, which I want to avoid and which is the reason why I am using Picasso. Also, it is not logical to be able to deviate from the default behavior in one direction, but not in the other.

My suggestion can be implemented in a few lines of code and is fully backward compatible, so I am asking you to reconsider. I believe this change will add value to Picasso.

**Comment by kevinvanmierlo** [comment]
Any update on this? I use a transform which blurs the image. If the image is taken from memory, it is taking too long and it instantly jumps to the new image. If I could force the fade, that would be great. This library already provides everything I need, except this.

**Comment by JakeWharton** [comment]
@kevinvanmierlo transformations are not applied to images from memory, they are delivered synchronously on the main thread at the time of request directly to the image view. This means that if you are, say, scrolling a list where all the images are available in memory you never see the loading fade.

**Comment by kevinvanmierlo** [comment]
@JakeWharton Ahh okay, stupid of me. Is there any way to do the transformation asynchronously? I really like that I can set a blurry background with such few lines of code.

**Comment by JakeWharton** [comment]
Transformations _only_ run in a background thread when the request is being fulfilled asynchronously. Either the blurred image was cached and can be used immediately or it has to be loaded and is done so on a background thread with the fade animation.

**Comment by kevinvanmierlo** [comment]
@JakeWharton Sorry for all my questions, but is there a way to force it to reload? When it is in cache it works great, and does everything instantly. From memory is the only one that takes long because of my transformation. If it cannot be done, I'll just use it without transformation.

**Comment by JakeWharton** [comment]
> but is there a way to force it to reload

Use `MemoryPolicy.NO_CACHE` on the request.

> From memory is the only one that takes long because of my transformation.

Transformations are not used when the image is loaded from memory. The cached image is already transformed. I don't know what you mean that it takes long because it's instant.

**Comment by kevinvanmierlo** [comment]
@JakeWharton Sorry, it seems I've been using the wrong word. Yeah, the cached image is instant. That works really great! But if I restart the app and Picasso doesn't load the image from network but from the disk (my bad). Then it does the transformation again to put it in the cache. When this happens it jumps to the new image instead of fading.

**Comment by JakeWharton** [comment]
That should not happen. If the image comes from anywhere except the memory cache the fade should be automatically applied. Are you using a `Target`?

**Comment by kevinvanmierlo** [comment]
@JakeWharton  I'm using this code:

```Java
Picasso.with(this)
                    .load(wallpaperUrl)
                    .placeholder(placeholder)
                    .error(placeholder)
                    .transform(new BlurTransformation(this))
                    .into(wallpaper);
```
Where wallpaperUrl is a String with an Url I want to load, placeholder an int using R.drawable and wallpaper my ImageView. I can provide the BlurTransformation as well if you'd like?

**Comment by kevinvanmierlo** [comment]
@JakeWharton I've created a test project to show this: https://github.com/kevinvanmierlo/PicassoBlurTest/tree/master When the image is on disk and Picasso loads it then it just jumps to the new image instead of fading.

**Comment by oliverhausler** [comment]
@kevinvanmierlo I believe the .transform() modifier is not suitable for animations. It's there for resizing and/or applying static filters, which modify the image only once, before it is cached, and which is not applied again during an image lifecycle [@JakeWharton correct me if I'm wrong]. Jake, may I suggest to provide a second command .transition() [or however you want to call it], which works like .transform(), but is applied every time the image is shown.

.transform() <-> .fit()/.centerCrop()/.center()
.transition() <-> .fade()/.noFace()

What I miss in Picasso is to be able to control the full pipeline of an image. It would be nice to have hooks for processing, caching, transitions, etc. Maybe something for Picasso Next?

**Comment by kevinvanmierlo** [comment]
@oliverhausler I didn't mean the transform() method should animate. Here's what I'm expecting: Load image --> perform transform --> when done fade from current image to loaded image with transform. It does so correctly when loading images from the web. It just doesn't fade when it's loading from disk. 

I do agree with you for more control. What I would like to see is being able to call the transform() method on the placeholder and error image (I'm using a background which should always be blurred, doesn't matter if it takes a second do blur)

**Comment by JakeWharton** [comment]
Picasso has no distinction between the web and the disk. They are the same place of origin and it will _always_ fade from that source (the HTTP client). It will only not fade when the image is cached in memory. I'll try to look at your sample later.

If you turn on debug indicators do you see green or blue when the image is loaded?

**Comment by kevinvanmierlo** [comment]
@JakeWharton Wow, I'm great at testing... Looking into this again it does fade. As expected it doesn't with memory. I'll delete the project. Thanks for the patience and the help!

**Comment by kevinvanmierlo** [comment]
@JakeWharton @oliverhausler Sorry for my problem between, I thought it was related to this issue. 

To continue with the issue is to have more control over the image. Here are two examples of what I would suggest for more control:
1. Force fade. For example when I want to load a new image if the user clicks on it and it is already in memory I would want it to fade.
2. Transform placeholder/error. For example when I have a background that always needs to be blurred it also needs to transform the placeholder or error holder.

See the last comment of @oliverhausler

**Comment by JakeWharton** [comment]
2 will never happen. The placeholder needs to be set synchronously on the main thread. If you want a blurred placeholder then you should pre-blur it to match. We _could_ transform the error drawable but I'm inclined not to for symmetry. If you need a blurred error image then it should be pre-blurred as well.

For 1 you can specify `noFade()` and then always play an animation in the image loaded callback. You can also rely on the callback being called synchronously to determine if an animation needs played.
```java
final AtomicBoolean playAnimation = new AtomicBoolean(true);
Picasso.with(context).load(..).into(iv, new Callback() {
  @Override public void onLoad() {
    if (playAnimation.get()) {
      // TODO play fade
    }
  }
  //..
});
playAnimation.set(false);

**Comment by kevinvanmierlo** [comment]
@JakeWharton Okay, I understand that 2 will never happen. I also understand that I could do this myself. If I see the source code it shouldn't be a lot of work to put forceFade in the functions. So I don't understand why it is a bad idea to suggest that? Correct me if I'm wrong. If it takes too much time I'm also happy to build it and apply a pull request if that's okay with you? Or do you really don't want it inside Picasso? @oliverhausler also correct me if I'm not saying what you intended here.

**Comment by ret702** [comment]
@kevinvanmierlo  HES THE BOSS NOW STFU

**Comment by GPack1** [comment]
@JakeWharton I dont understand the AtomicBoolean use. Why setting false immediatly after calling Picasso?

**Comment by realdeadfish** [comment]
@GPack1 because

> For 1 you can specify noFade() and then always play an animation in the image loaded callback. You can also rely on the callback being called synchronously to determine if an animation needs played.

I.e. if you load from memory, there is no thread-switching involved, so the whole picasso pipe including `.get()` is executed before `playAnimation` is set to `false`.

**Comment by mhenry** [comment]
If you're scratching your head trying to figure out why @JakeWharton's solution isn't always working the way you'd expect, I've found that it helps to post a `Runnable` to set the `AtomicBoolean`. That helps ensure the callback has had a chance to execute.

```java
iv.post(new Runnable() {
    @Override
    public void run() {
        playAnimation.set(false);
    }
});
```
